.pre-commit-config.yaml

```diff
@@ -0,0 +1,206 @@
+---
+# Pre-commit hooks definition file
+#
+# This file holds information for file formatting and validation. See
+# documentation here:
+# https://numpy.org/doc/stable/dev/development_workflow.html?highlight=development%20workflow
+# Or the pre-commit docs here:
+# https://numpy.org/doc/stable/dev/development_workflow.html
+#
+# If you add or adjust hooks, or want to update to the latest version, use
+# `pre-commit autoupdate --freeze`. This pins the version to the commit hash
+# rather than a (movable) tag, which adds some security benefit. This sets the
+# "rev:" key and the adds the "# frozen: vx.x.x" comments that you see.
+#
+# Note that additional_dependencies pins need to be manually updated. Try to
+# ensure versiong stay in sync when a dependency is used across multiple
+# hooks.
+
+repos:
+  ### Always-run hooks ###
+  # This section includes hooks that run whenever relevant files change.
+  # Basic checkers, fairly self-explanatory
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: 3298ddab3c13dd77d6ce1fc0baf97691430d84b0  # frozen: v4.3.0
+    hooks:
+      # File validity checkers
+      - id: check-yaml
+      - id: check-toml
+
+      # Simple cleanliness things
+      - id: fix-byte-order-marker
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+      - id: mixed-line-ending
+      # Defaults to 500 kB
+      - id: check-added-large-files
```

Comment:
Why do we need this?

Reply:
These three checks are basically instant and come for free with pre-commit. The whitespace and line endings are pretty common things that invisibly sneak into commits, e.g. from pep8:

> Avoid trailing whitespace anywhere. Because it’s usually invisible, it can be confusing: e.g. a backslash followed by a space and a newline does not count as a line continuation marker. Some editors don’t preserve it and many projects (like CPython itself) have pre-commit hooks that reject it

The large files check isn't really necessary because of the squash & merge workflow, but does prevent somebody from accidentally committing something like an array dump file in the first place. And it's an instant check so no harm in having it.